docs/DesignDocs/ExtendedEnclaveInitializationData.md

```diff
@@ -0,0 +1,156 @@
+        props.header.size_settings.num_stack_pages =
+            eeid->size_settings.num_stack_pages;
+        props.header.size_settings.num_tcs = eeid->size_settings.num_tcs;
+
+        // Patch ELF symbols
+        elf64_sym_t sym_props = {0};
+        const elf64_t* eimg = &oeimage.u.elf.elf;
+        if (elf64_find_symbol_by_name(
+                eimg, "oe_enclave_properties_sgx", &sym_props) == 0)
+        {
+            uint64_t* sym_props_addr = NULL;
+            sym_props_addr =
+                (uint64_t*)(oeimage.image_base + sym_props.st_value);
+            oe_sgx_enclave_properties_t* p =
+                (oe_sgx_enclave_properties_t*)sym_props_addr;
+            p->header.size_settings = props.header.size_settings;
+        }
+    }
+```
+
+Then, after the enclave image is loaded but before it is initialized, we add the additional memory pages containing the contents of the `oe_eeid_t` struct:
+
+```C
+        oe_sha256_context_t* hctx = &context->hash_context;
+        sgx_sigstruct_t* sigstruct = (sgx_sigstruct_t*)properties->sigstruct;
+        memcpy(eeid->sigstruct, (uint8_t*)sigstruct, sizeof(sgx_sigstruct_t));
+        eeid->vaddr = *vaddr;
+        eeid->entry_point = oeimage.entry_rva;
+
+        uint64_t ee_sz = sizeof(oe_eeid_t) + eeid->data_size;
+        uint64_t epg_sz = eeid_pages_size(eeid);
+        uint64_t num_pages = epg_sz / OE_PAGE_SIZE;
+        assert(*vaddr == enclave_end - epg_sz);
+
+        oe_page_t* pages = (oe_page_t*)eeid;
+        if (ee_sz < epg_sz)
+        {
+            oe_page_t* tmp = (oe_page_t*)calloc(1, epg_sz);
+            memcpy(tmp, eeid, ee_sz);
+            pages = tmp;
+        }
+
+        OE_CHECK(_add_extra_data_pages(
+            context, enclave->addr, pages, num_pages, vaddr));
+
+        if (ee_sz < epg_sz)
+            free(pages);
+
+        OE_SHA256 ext_mrenclave;
+        oe_sha256_final(hctx, &ext_mrenclave);
+```
+
+Finally, since our changes invalidate the signature of the base image, we need to dynamically re-sign the image. For this, we use the `OE_DEBUG_SIGN_KEY`, which is well-known by all OE users and thus does not require additional public-key infrastructure support. Note that applications should never rely on the apparent signing entity (e.g. `MRSIGNER`) of EEID-enabled evidence. Instead, what we provide to the application when we appraise EEID evidence, is the evidence of the base image.
+
+```C
+        OE_CHECK(oe_sgx_sign_enclave(
+            &ext_mrenclave,
+            properties->config.attributes,
+            properties->config.product_id,
+            properties->config.security_version,
+            OE_DEBUG_SIGN_KEY,
+            OE_DEBUG_SIGN_KEY_SIZE,
+            sigstruct));
+```
+
+We also add a serialization function `oe_serialize_eeid` for the `oe_eeid_t`, so that EEID can be sent to verifiers as an endorsement for verification of the base image evidence together with the EEID configuration of the instance.
```

Comment:
From an earlier unaddressed comment:
> This is problematic as it tightly couples a specific attester/verifier implementation to a general public API provided by the OE SDK. Do you expect that those are also ingested as part of the OE SDK?
>
> Specifically, it feels like the OE SDK is modified to privilege a specific use case as inherent in the OE API even though it's not generalizable beyond SGX1; In SGX2 with EDMM and KSS support, the problems that this design attempts to solve are natively supported. It's not clear to me how this design migrates forward to SGX2, or generalizes to other TEE, which is a core architectural premise of the OE SDK.

This still holds true for the updated design; I think that at a higher level, it's inevitable that OE will move a lot of the SGX1-specific semantics out of its API surface into a TEE extension model; at that point, this will likely be factored out into the TEE extension. In the meantime, it's useful to explore the use cases for something like this, and I would be in favor of having this as an [experimental API](https://github.com/openenclave/openenclave/blob/d7c8e614221a2ad586b2a4eae144b1672f636e9a/docs/ApiGuidelines.md#experimental-apis), with the expectation that it will likely evolve further into a different API.

Reply:
I like the idea of "having this be an experimental API with the expectation that it will likely evolve further", and that would let this be merged now.

Reply:
I totally agree, this is clearly experimental, and I can see that changes will likely be necessary for non-SGX platforms (e.g. at least a replacement for sigstruct). I'll check with the team, but I don't expect any objections.

Reply:
Forgot to address the SGX2 comment before: SGX2's config register does not solve the same problem, as it provides a slightly weaker security guarantee. As in our example of a script running in an interpreter within the enclave, suppose that interpreter and script come from different parties, but they are now running in the same (enclave) memory. The config register does not address the problem of malicious scripts that have access to `oe_get_report` (and similar) which they can use to fake evidence, even if the config register matches.